lib/Dialect/Torch/Transforms/RecomposeComplexOps.cpp

```diff
@@ -110,6 +110,48 @@ class RecomposeSliceCopy_ : public OpRewritePattern<AtenCopy_Op> {
   }
 };
 
+class RecomposeSelectCopy_ : public OpRewritePattern<AtenCopy_Op> {
+public:
+  using OpRewritePattern::OpRewritePattern;
+  LogicalResult matchAndRewrite(AtenCopy_Op op,
+                                PatternRewriter &rewriter) const override {
+    // Similar to RecomposeSliceCopy_, here is also to prevent mutation
+    if (!op.use_empty())
+      return rewriter.notifyMatchFailure(
+          op, "`AtenCopy_Op` must not have any users");
+    if (!op.getSelf().getDefiningOp() ||
+        !isa<AtenSelectIntOp>(op.getSelf().getDefiningOp()))
+      return rewriter.notifyMatchFailure(
+          op, "defining op is not `AtenSelectIntOp`");
+    auto selectOp = cast<AtenSelectIntOp>(op.getSelf().getDefiningOp());
+
+    // Get indices
+    int64_t dim;
+    if (!matchPattern(selectOp.getDim(), m_TorchConstantInt(&dim)))
+      return failure();
+    int64_t index;
+    if (!matchPattern(selectOp.getIndex(), m_TorchConstantInt(&index)))
+      return failure();
+
+    Value oneVal = rewriter.create<ConstantIntOp>(op.getLoc(), 1);
+    Value newStart = selectOp.getIndex();
+    Value dimSize = rewriter.create<AtenSizeIntOp>(
+        op.getLoc(), selectOp.getSelf(), selectOp.getDim());
+    if (index < 0) {
+      newStart = rewriter.create<AtenAddIntOp>(
+        op.getLoc(), newStart, dimSize);
+    }
+    Value newEnd = rewriter.create<AtenAddIntOp>(op.getLoc(), newStart, oneVal);
+
+    Value selectOpInput = selectOp.getSelf();
+    rewriter.replaceOpWithNewOp<AtenSliceTensorOp>(
```

Comment:
I think we need to be careful about the way shape information is handled. In the decomposition of `AtenSelectIntOp` into `AtenSliceTensorOp` an extra squeeze is needed:

https://github.com/llvm/torch-mlir/blob/598d6907108316aa050816c4a1fe067ae07ba736/lib/Dialect/Torch/Transforms/DecomposeComplexOps.cpp#L308-L310

Can you test the implementation with an MLIR snippet like the one in your PR description but with shapes in all the `!torch.tensor` types?

Reply:
# Change Description
I've added two unit tests: 
- for select + copy_ between same type in int64, 
- for select + copy_ between int64 and float32 types.
# Question
Could I trouble you with a couple of questions I'd like to clarify, please:

1. I noticed there are 13 types of torch.tensor, but most tests only cover float32 and int64. Should I consider adding other types as well? 
2. Regarding the second test case mentioned earlier, it resulted in an XFAIL. Is it okay if I add it to `COMMON_TORCH_MLIR_LOWERING_XFAILS`?

# About extra squeeze
From my understanding, `RecomposeSelectCopy_` won't affect individual select operators; its impact is limited to the select + copy_ combination operator. Once the select is transformed into a slice operator, it will be handled by `RecomposeSliceCopy_`. Therefore, it seems there's no need to add any extra squeeze operator.

Reply:
> I noticed there are 13 types of torch.tensor, but most tests only cover float32 and int64. Should I consider adding other types as well?

No need to test every type. We usually test just the most commonly used types. Unless your pattern has special logic for a particular type, you should also test that, but if your pattern is independent of type, doing float32 or int64 is enough.

> Regarding the second test case mentioned earlier, it resulted in an XFAIL. Is it okay if I add it to COMMON_TORCH_MLIR_LOWERING_XFAILS?

We try to avoid adding tests to `COMMON_TORCH_MLIR_LOWERING_XFAILS`. A test that does not work on any backend is concerning. What error are you seeing?

> About extra squeeze

It does seem to affect it.

```
module {
  func.func @forward(%arg0: !torch.vtensor<[100,100],f32>, %arg1: !torch.vtensor<[100],f32>) -> (!torch.tensor<[100,100],f32>, !torch.tensor<[100],f32>) {
    %int0 = torch.constant.int 0
    %false = torch.constant.bool false
    %0 = torch.copy.to_tensor %arg0 : !torch.tensor<[100,100],f32>
    %1 = torch.copy.to_tensor %arg1 : !torch.tensor<[100],f32>
    %2 = torch.aten.select.int %0, %int0, %int0 : !torch.tensor<[100,100],f32>, !torch.int, !torch.int -> !torch.tensor<[100],f32>
    %3 = torch.aten.copy_ %2, %1, %false : !torch.tensor<[100],f32>, !torch.tensor<[100],f32>, !torch.bool -> !torch.tensor<[100],f32>
    return %0, %2 : !torch.tensor<[100,100],f32>, !torch.tensor<[100],f32>
  }
}
```

If you run the following command on the above IR, you get an error:

```
~/torch-mlir/build/bin/torch-mlir-opt ~/mlir-files/temp.mlir --torch-recompose-complex-ops --mlir-print-ir-after-failure
```

```
/usr/local/google/home/ramiroleal/mlir-files/temp.mlir:9:3: error: type of return operand 1 ('!torch.tensor<[100,100],f32>') doesn't match function result type ('!torch.tensor<[100],f32>') in function @forward
  return %1, %4 : !torch.tensor<[100,100],f32>, !torch.tensor<[100],f32>
  ^
/usr/local/google/home/ramiroleal/mlir-files/temp.mlir:9:3: note: see current operation: "func.return"(%1, %1) : (!torch.tensor<[100,100],f32>, !torch.tensor<[100,100],f32>) -> ()
// -----// IR Dump After RecomposeComplexOps Failed (torch-recompose-complex-ops) //----- //
"func.func"() <{function_type = (!torch.vtensor<[100,100],f32>, !torch.vtensor<[100],f32>) -> (!torch.tensor<[100,100],f32>, !torch.tensor<[100],f32>), sym_name = "forward"}> ({
^bb0(%arg0: !torch.vtensor<[100,100],f32>, %arg1: !torch.vtensor<[100],f32>):
  %0 = "torch.constant.bool"() {value = false} : () -> !torch.bool
  %1 = "torch.copy.to_tensor"(%arg0) : (!torch.vtensor<[100,100],f32>) -> !torch.tensor<[100,100],f32>
  %2 = "torch.copy.to_tensor"(%arg1) : (!torch.vtensor<[100],f32>) -> !torch.tensor<[100],f32>
  %3 = "torch.aten.copy_"(%1, %2, %0) : (!torch.tensor<[100,100],f32>, !torch.tensor<[100],f32>, !torch.bool) -> !torch.tensor<[100],f32>
  "func.return"(%1, %1) : (!torch.tensor<[100,100],f32>, !torch.tensor<[100,100],f32>) -> ()
}) : () -> ()
```

Reply:
# Change Description
I truly appreciate your thorough response. Indeed, your view is accurate. I realize that I had a misunderstanding in this particular aspect. I have made the following adjustments:
- I made modifications to the implementation of `RecomposeSelectCopy_` to enable it to handle situations where the select operator alters the shape of the tensor. 
- I have moved the unit test named SelectIntCopyFloat to the LINALG_XFAIL_SET section. 
- Within the implementation of `RecomposeSelectCopy_`, I reuse the `computeReductionType` function. Consequently, I have transitioned it from a static function to a global one and relocated it to the `torch-mlir/Dialect/Torch/Utils/Utils directory`. 

Your valuable insights and suggestions in these matters are sincerely appreciated.
# Test
Both cases are now handled correctly wit this command:
```
~/torch-mlir/build/bin/torch-mlir-opt ~/mlir-files/temp.mlir --torch-recompose-complex-ops --mlir-print-ir-after-failure
```
Not remove `torch.aten.select.int`
```
module {
  func.func @forward(%arg0: !torch.vtensor<[100,100],f32>, %arg1: !torch.vtensor<[100],f32>) -> (!torch.tensor<[100,100],f32>, !torch.tensor<[100],f32>) {
    %int0 = torch.constant.int 0
    %false = torch.constant.bool false
    %0 = torch.copy.to_tensor %arg0 : !torch.tensor<[100,100],f32>
    %1 = torch.copy.to_tensor %arg1 : !torch.tensor<[100],f32>
    %2 = torch.aten.select.int %0, %int0, %int0 : !torch.tensor<[100,100],f32>, !torch.int, !torch.int -> !torch.tensor<[100],f32>
    %3 = torch.aten.copy_ %2, %1, %false : !torch.tensor<[100],f32>, !torch.tensor<[100],f32>, !torch.bool -> !torch.tensor<[100],f32>
    return %0, %2 : !torch.tensor<[100,100],f32>, !torch.tensor<[100],f32>
  }
}
```
```
module {
  func.func @forward(%arg0: !torch.vtensor<[100,100],f32>, %arg1: !torch.vtensor<[100],f32>) -> (!torch.tensor<[100,100],f32>, !torch.tensor<[100],f32>) {
    %int1 = torch.constant.int 1
    %int0 = torch.constant.int 0
    %none = torch.constant.none
    %false = torch.constant.bool false
    %0 = torch.copy.to_tensor %arg0 : !torch.tensor<[100,100],f32>
    %1 = torch.copy.to_tensor %arg1 : !torch.tensor<[100],f32>
    %2 = torch.aten.select.int %0, %int0, %int0 : !torch.tensor<[100,100],f32>, !torch.int, !torch.int -> !torch.tensor<[100],f32>
    %3 = torch.aten.arange.start_step %int0, %int1, %int1, %none, %none, %none, %none : !torch.int, !torch.int, !torch.int, !torch.none, !torch.none, !torch.none, !torch.none -> !torch.tensor<[?],f32>
    %4 = torch.prim.ListConstruct %3 : (!torch.tensor<[?],f32>) -> !torch.list<optional<tensor>>
    %5 = torch.aten._index_put_impl_ %0, %4, %1, %false, %false : !torch.tensor<[100,100],f32>, !torch.list<optional<tensor>>, !torch.tensor<[100],f32>, !torch.bool, !torch.bool -> !torch.tensor<[100,100],f32>
    return %0, %2 : !torch.tensor<[100,100],f32>, !torch.tensor<[100],f32>
  }
}
```
Remove `torch.aten.select.int`
```
module {
  func.func @forward(%arg0: !torch.vtensor<[100,100],f32>, %arg1: !torch.vtensor<[100],f32>) -> (!torch.tensor<[100,100],f32>) {
    %int0 = torch.constant.int 0
    %false = torch.constant.bool false
    %0 = torch.copy.to_tensor %arg0 : !torch.tensor<[100,100],f32>
    %1 = torch.copy.to_tensor %arg1 : !torch.tensor<[100],f32>
    %2 = torch.aten.select.int %0, %int0, %int0 : !torch.tensor<[100,100],f32>, !torch.int, !torch.int -> !torch.tensor<[100],f32>
    %3 = torch.aten.copy_ %2, %1, %false : !torch.tensor<[100],f32>, !torch.tensor<[100],f32>, !torch.bool -> !torch.tensor<[100],f32>
    return %0 : !torch.tensor<[100,100],f32>
  }
}
```
```
module {
  func.func @forward(%arg0: !torch.vtensor<[100,100],f32>, %arg1: !torch.vtensor<[100],f32>) -> !torch.tensor<[100,100],f32> {
    %int1 = torch.constant.int 1
    %int0 = torch.constant.int 0
    %none = torch.constant.none
    %false = torch.constant.bool false
    %0 = torch.copy.to_tensor %arg0 : !torch.tensor<[100,100],f32>
    %1 = torch.copy.to_tensor %arg1 : !torch.tensor<[100],f32>
    %2 = torch.aten.arange.start_step %int0, %int1, %int1, %none, %none, %none, %none : !torch.int, !torch.int, !torch.int, !torch.none, !torch.none, !torch.none, !torch.none -> !torch.tensor<[?],f32>
    %3 = torch.prim.ListConstruct %2 : (!torch.tensor<[?],f32>) -> !torch.list<optional<tensor>>
    %4 = torch.aten._index_put_impl_ %0, %3, %1, %false, %false : !torch.tensor<[100,100],f32>, !torch.list<optional<tensor>>, !torch.tensor<[100],f32>, !torch.bool, !torch.bool -> !torch.tensor<[100,100],f32>
    return %0 : !torch.tensor<[100,100],f32>
  }
}
```